File: Problems 0200-0299/297. Serialize and Deserialize Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {
    // // BFS
    // public String serialize(TreeNode root) {
    //     Queue<TreeNode> q = new LinkedList<>();
    //     q.offer(root);

    //     StringBuilder sb = new StringBuilder();
    //     while (!q.isEmpty()) {
    //         TreeNode node = q.poll();

    //         if (node == null) {
    //             sb.append("# ");
    //         } else {
    //             sb.append(node.val + " ");
    //             q.offer(node.left);
    //             q.offer(node.right);
    //         }
    //     }

    //     return sb.toString().trim();
    // }

    // public TreeNode deserialize(String data) {
    //     String[] nodes = data.split(" ");
    //     if (nodes[0].equals("#")) {
    //         return null;
    //     }
    //     TreeNode root = new TreeNode(Integer.valueOf(nodes[0]));
    //     Queue<TreeNode> q = new LinkedList<>();
    //     q.offer(root);

    //     for (int i = 1; i < nodes.length; i++) {
    //         TreeNode node = q.poll();
    //         if (!nodes[i].equals("#")) {
    //             TreeNode left = new TreeNode(Integer.valueOf(nodes[i]));
    //             node.left = left;
    //             q.offer(left);
    //         }
    //         if (!nodes[++i].equals("#")) {
    //             TreeNode right = new TreeNode(Integer.valueOf(nodes[i]));
    //             node.right = right;
    //             q.offer(right);
    //         }
    //     }
    //     return root;
    // }

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        // List<String> data = new ArrayList<>();
        // dfsSerialize(root, data);
        // return String.join(" ", data);
        StringBuilder sb = new StringBuilder();
        dfsSerialize(root, sb);
        return sb.toString().trim();
    }

    private void dfsSerialize(TreeNode node, StringBuilder data) {
        if (node == null) {
            data.append("# ");
        } else {
            data.append(node.val + " ");
            dfsSerialize(node.left, data);
            dfsSerialize(node.right, data);
        }
    }

    private void dfsSerialize(TreeNode node, List<String> data) {
        if (node == null) {
            data.add("#");
        } else {
            data.add(node.val + "");
            dfsSerialize(node.left, data);
            dfsSerialize(node.right, data);
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] parts = data.split(" ");

        return dfsDeserialize(Arrays.asList(parts).iterator());
        // return dfsDeserialize(new LinkedList<>(Arrays.asList(parts)));
    }

    private TreeNode dfsDeserialize(Queue<String> tree) {
        String val = tree.poll();
        if (val.equals("#")) {
            return null;
        }
        TreeNode node = new TreeNode(Integer.valueOf(val));
        node.left = dfsDeserialize(tree);
        node.right = dfsDeserialize(tree);
        return node;
    }

    private TreeNode dfsDeserialize(Iterator<String> tree) {
        String val = tree.next();
        if (val.equals("#")) {
            return null;
        }
        TreeNode node = new TreeNode(Integer.valueOf(val));
        node.left = dfsDeserialize(tree);
        node.right = dfsDeserialize(tree);
        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: Problems 0200-0299/297. Serialize and Deserialize Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# DFS
# class Codec:

#     def serialize(self, root):
#         """Encodes a tree to a single string.
        
#         :type root: TreeNode
#         :rtype: str
#         """
        
#         def dfs(node, encode):
#             if not node:
#                 encode.append('#')
#                 return
            
#             encode.append(str(node.val))
#             dfs(node.left, encode)
#             dfs(node.right, encode)
        
#         encode = []
#         dfs(root, encode)
#         return ",".join(encode)
        

#     def deserialize(self, data):
#         """Decodes your encoded data to tree.
        
#         :type data: str
#         :rtype: TreeNode
#         """
#         # if data == "#":
#         #     return None
        
#         # nodes = data.split(',')
#         # root = TreeNode(int(nodes[0]))

#         # stack = [root]

#         # i = 1

#         # while stack:
#         #     # If we're processing a node here, that means not
#         #     # all of its children has been processed yet
#         #     curr = stack[-1]

#         #     if curr.left is None:
#         #         val = nodes[i]
#         #         i += 1
#         #         # Add left child and continue (since we're processing
#         #         # all left childs first)
#         #         if val != '#':
#         #             curr.left = TreeNode(int(val))
#         #             stack.append(curr.left)
#         #             continue

#         #     # If code reaches here that means curr.left is filled,
#         #     # so we need to fill the right child now
#         #     # And since we're filling the right child, we've finished
#         #     # processing this node, so we can pop it off the stack
#         #     stack.pop()
            
#         #     # if curr.right is None:
#         #     val = nodes[i]
#         #     i += 1
#         #     if val != '#':
#         #         curr.right = TreeNode(int(val))
#         #         stack.append(curr.right)
        
#         # return root

#         if data == "#":
#             return None
        
#         nodes = data.split(',')
#         nodes = iter(nodes)

#         def dfs(nodes):
#             val = next(nodes)

#             if val == '#':
#                 return None
            
#             node = TreeNode(int(val))
#             node.left = dfs(nodes)
#             node.right = dfs(nodes)
#             return node

#         return dfs(nodes)

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "#"
        
        q = deque([root])
        res = []
        while q:
            node = q.popleft()

            if node:
                res.append(str(node.val))
                q.append(node.left)
                q.append(node.right)
            else:
                res.append('#')

        return ",".join(res)

        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == "#":
            return None
        
        nodes = data.split(',')
        root = TreeNode(int(nodes[0]))
        q = deque([root])
        i = 1
        
        while i < len(nodes):
            node = q.popleft()
            if nodes[i] != '#':
                node.left = TreeNode(int(nodes[i]))
                q.append(node.left)
            
            i += 1
            if nodes[i] != '#':
                node.right = TreeNode(int(nodes[i]))
                q.append(node.right)
            
            i += 1
        
        return root

            
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))